event/store.go

```diff
@@ -0,0 +1,60 @@
+package event
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+const defaultEventCountLimit = 5
+
+// Store stores events
+type Store struct {
+	*sync.RWMutex
+
+	events map[string][]*Event // taskname => events
+	limit  int
+}
+
+// NewStore returns a new store
+func NewStore() *Store {
+	return &Store{
+		RWMutex: &sync.RWMutex{},
+		events:  make(map[string][]*Event),
+		limit:   defaultEventCountLimit,
+	}
+}
+
+// Add adds an event and manages the limit of number of events stored per task.
+func (s *Store) Add(e *Event) error {
+	if e == nil {
+		return errors.New("error adding event: nil event")
+	}
+	if e.TaskName == "" {
+		return fmt.Errorf("error adding event: taskname cannot be empty %s", e.GoString())
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	events := s.events[e.TaskName]
+	events = append([]*Event{e}, events...) // prepend
+	if len(events) > s.limit {
+		events = events[:len(events)-1]
+	}
+	s.events[e.TaskName] = events
```

Comment:
Go lists are by reference and modifying `events` variable updates the list within the map too ([example](https://play.golang.com/p/ZbtITuQo-3t)). No need to change but seeing this thought i'd mention it

Reply:
Ah yes! I tried removing this line and discovered that `append` creates and returns a new slice so events needs to be reassigned. I'll also keep in mind though that events is a reference, which I had forgotten!

Reply:
oh true on the `append` 👍